R/env.R

```diff
@@ -311,176 +367,220 @@ mut_parent_env <- function(env, new_env) {
+#' # fn() now sees the objects:
 #' fn()
-#'
-#' # Redefine new bindings:
-#' fn <- env_assign(fn, "a", "1")
-#' fn <- env_define(fn, b = "2", c = "3")
-#' fn()
-env_assign <- function(env = caller_env(), nm, x) {
-  env_ <- get_env(env)
-  base::assign(nm, x, envir = env_)
-  env
-}
-#' @rdname env_assign
-#' @export
-env_bind <- function(env = caller_env(), data = list()) {
-  stopifnot(is_dictionaryish(data))
+env_bind <- function(.env, ...) {
+  data <- dots_list(...)
+  stopifnot(is_named(data))
+
   nms <- names(data)
+  env_ <- get_env(.env)
 
-  env_ <- get_env(env)
   for (i in seq_along(data)) {
     nm <- nms[[i]]
     base::assign(nm, data[[nm]], envir = env_)
   }
 
-  env
+  .env
 }
-#' @rdname env_assign
-#' @export
-env_define <- function(env = caller_env(), ...) {
-  env_bind(env, list(...))
-}
-
-#' Assign a promise to an environment.
-#'
-#' These functions let you create a promise in an environment. Such
-#' promises behave just like lazily evaluated arguments. They are
-#' evaluated whenever they are touched by code, but not when they are
-#' passed as arguments.
-#'
-#' @inheritParams env_assign
-#' @param expr An expression to capture for
-#'   `env_assign_promise()`, or a captured expression (either
-#'   quoted or a formula) for the standard evaluation version
-#'   `env_assign_promise_()`. This expression is used to create a
-#'   promise in `env`.
-#' @param eval_env The environment where the promise will be evaluated
-#'   when the promise gets forced. If `expr` is a formula, its
-#'   environment is used instead. If not a formula and `eval_env` is
-#'   not supplied, the promise is evaluated in the environment where
-#'   `env_assign_promise()` (or the underscore version) was called.
-#' @seealso [env_assign()], [env_assign_active()]
+#' @rdname env_bind
+#' @param .eval_env The environment where the expressions will be
+#'   evaluated when the symbols are forced.
 #' @export
 #' @examples
-#' env <- child_env()
-#' env_assign_promise(env, "name", cat("forced!\n"))
+#'
+#' # env_bind_exprs() assigns expressions lazily:
+#' env <- env()
+#' env_bind_exprs(env, name = cat("forced!\n"))
+#' env$name
 #' env$name
 #'
-#' # Use the standard evaluation version with quoted expressions:
-#' f <- ~message("forced!")
-#' env_assign_promise_(env, "name2", f)
-#' env$name2
-env_assign_promise <- function(env = caller_env(), nm, expr,
-                               eval_env = caller_env()) {
-  f <- as_quosure(substitute(expr), eval_env)
-  env_assign_promise_(env, nm, f)
-}
-#' @rdname env_assign_promise
-#' @export
-env_assign_promise_ <- function(env = caller_env(), nm, expr,
-                                eval_env = caller_env()) {
-  args <- list(
-    x = nm,
-    value = get_expr(expr),
-    eval.env = get_env(expr, eval_env),
-    assign.env = get_env(env)
-  )
-  do.call("delayedAssign", args)
-}
+#' # You can unquote expressions. Note that quosures are not
+#' # supported, only raw expressions:
```

Comment:
We're capturing the expression. If we say we support quosures, that means we allow users to unquote quosures into the captured expression. Unless we evaluate the promise in an overscope, that will only work at top-level and therefore will probably be confusing.